File: others/problems/search/divide-conquer/median.py
#!/usr/bin/python

# Find the median of two sorted arrays.
# Recursive solution

# For the two arrays A and B:

# General recursive case:
#   Calculate the medians m1 and m2 of the input arrays A and B respectively.
#   If m1 < m2:
#      Recursive find the median of two sub-arrays: A[m1...] and B[...m2]
#   Otherwise, m2 < m1:
#      Recursive find the median of two sub-arrays: A[...m1] and B[m2...]

import random

def median(xs, ys):
    n, m = len(xs), len(ys)
    med = kth(xs, 0, n, ys, 0, m, (m + n) / 2 + 1)
    if (n + m) % 2 == 0:
        return (med + kth(xs, 0, n, ys, 0, m, (m + n) / 2)) / 2.0
    return med

# find the k-th element from xs[x0, x1) and ys[y0, y1), k starts from 1
def kth(xs, x0, x1, ys, y0, y1, k):
    if x1 - x0 < y1 - y0:
        return kth(ys, y0, y1, xs, x0, x1, k)
    if x1 <= x0:
        return ys[y0 + k - 1]
    if y1 <= y0:
        return xs[x0 + k - 1]
    if k == 1:
        return min(xs[x0], ys[y0])
    j = min(k/2, y1 - y0)
    i = k - j
    i, j = x0 + i, y0 + j
    if xs[i-1] < ys[j-1]:
        return kth(xs, i, x1, ys, y0, j, k - i + x0)
    else:
        return kth(xs, x0, i, ys, j, y1, k - j + y0)

N = 100

def test():
    for _ in range(N):
        xs = range(random.randint(1, N))
        m = (len(xs) - 1) / 2.0
        random.shuffle(xs)
        n = random.randint(0, len(xs))
        (xs, ys) = (sorted(xs[:n]), sorted(xs[n:]))
        m1 = median(xs, ys)
        if abs(m - m1) > 0.005:
            print "FAIL: median expected:", m, "actual:", m1
            print xs, "len=", len(xs)
            print ys, "len=", len(ys)
            exit()

if __name__ == "__main__":
    test()

File: others/problems/search/divide-conquer/median.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>

#define N 10
#define min(a, b) (a < b ? a : b)

int kth(int A[], int m, int B[], int n, int k) {
    int i, j;
    if (m < n) return kth(B, n, A, m, k);
    if (n == 0) return A[k - 1];
    if (k == 1) return min(A[0], B[0]);
    j = min(k/2, n);
    i = k - j;
    return A[i - 1] < B[j - 1] ?
        kth(A + i, m - i, B, j, k - i):
        kth(A, i, B + j, n - j, k - j);
}

double median(int A[], int m, int B[], int n) {
    int k = (m + n) / 2;
    double x = (double) kth(A, m, B, n, k + 1);
    if (!((m + n) & 0x1)) /*odd*/
        x = (x + (double) kth(A, m, B, n, k) / 2.0;
    return x;
}

/*for verification*/
int cmp(const void* x, const void* y) {
    return *(const int*)x - *(const int*)y;
}

void swap(int *a, int *b) {
    int tmp = *a; *a = *b; *b = tmp;
}

void print(int *xs, int n) {
    int i;
    printf("[");
    for (i = 0; i < n; ++i)
        printf(i == n-1 ? "%d]\n" : "%d, ", xs[i]);
}

void check(int *xs, int n, int* ys, int m, double k) {
    double med = median(xs, n, ys, m);
    if (fabs(med - k) > 0.005) {
        printf("FAIL: expected: %f, actual: %f\n", k, med);
        print(xs, n);
        print(ys, m);
        exit(-1);
    }
}

int main(int argc, char** argv) {
    int i, j, n, m, c = N;
    int xs[N], ys[N];
    double k;
    while(--c) {
        n = rand() % N + 1;
        for (i = 0; i < n; ++i)
            xs[i] = i;
        k = (double)(n - 1) / 2.0;
        for (i = 0; i < n; ++i) {
            j = rand() % n;
            swap(&xs[i], &xs[j]);
        }
        m = rand() % n;
        for (i = 0; i < m; ++i)
            ys[i] = xs[n - 1 - i];
        n -= m;
        qsort(xs, n, sizeof(int), cmp);
        qsort(ys, m, sizeof(int), cmp);
        check(xs, n, ys, m, k);
    }
    return 0;
}
